ModulePresentationsForCAP/gap/ModulePresentationsForCAP.gi

```diff
@@ -2146,3 +2172,296 @@ end );
+
+  function( category )
+
+    AddRandomObject( category,
+        function( category, n )
+        local homalg_ring, left_presentations, M;
+
+        homalg_ring := category!.ring_for_representation_category;
+
+        left_presentations := LeftPresentations( homalg_ring );
+
+        M := RandomObject( left_presentations, n );
+
+        return AsRightPresentation( Involution( UnderlyingMatrix( M ) ) );
+
+      end );
+
+end );
+
+##
+## Interpretation of n:
+## The number of relations of the range of the created random morphism is n.
+InstallGlobalFunction( ADD_RANDOM_MORPHISM_WITH_FIXED_SOURCE_LEFT,
+    function( category )
+      local homalg_ring;
+
+      homalg_ring := category!.ring_for_representation_category;
+
+      AddRandomMorphismWithFixedSource( category,
+        function( M, n )
+          local m, y, syz, x, u, U, e;
+          
+          if n = 0 then
+            
+            return UniversalMorphismIntoZeroObject( M );
+
+          fi;
+
+          m := UnderlyingMatrix( M );
+
+          y := homalg_ring!.random_matrix_func( NrRows( m ), n );
+
+          syz := SyzygiesOfColumns( UnionOfColumns( m, y ) );
+
+          if NrCols( syz ) > 0 then
+
+            e := homalg_ring!.random_element_func( );
+
+            e := Random( [ e, e, e, One( homalg_ring ) ] );
+
+            syz := e * CertainColumns( syz, [ 1 .. Random( [ 1 .. NrCols( syz ) ]  ) ] );
+
+          fi;
+
+          x := CertainRows( syz, [ 1 .. NrCols( m ) ] );
+
+          u := CertainRows( syz, [ NrCols( m ) + 1 .. NrRows( syz ) ] );
+
+          U := AsLeftPresentation( u );
+
+          return PresentationMorphism( M, x, U );
+
+    end );
+
+end );
+
+##
+## Interpretation of n:
+## The number of generators of the source of the created random morphism is n.
+InstallGlobalFunction( ADD_RANDOM_MORPHISM_WITH_FIXED_RANGE_LEFT,
+    function( category )
+      local homalg_ring;
+
+      homalg_ring := category!.ring_for_representation_category;
+
+      AddRandomMorphismWithFixedRange( category,
+        function( U, n )
+          local u, x, syz, m, M, e;
+          
+          if n = 0 then
+            
+            return UniversalMorphismFromZeroObject( U );
+
+          fi;
+
+          u := UnderlyingMatrix( U );
+
+          x := homalg_ring!.random_matrix_func( n, NrCols( u ) );
+
+          syz := SyzygiesOfRows( UnionOfRows( x, u ) );
+
+          if NrRows( syz ) > 0 then
+
+            e := homalg_ring!.random_element_func( );
+
+            e := Random( [ e, e, e, e, One( homalg_ring ) ] );
+
+            syz := e * CertainRows( syz, [ 1 .. Random( [ 1 .. NrRows( syz ) ] ) ] );
+
+          fi;
+
+          m := CertainColumns( syz, [ 1 .. NrRows( x ) ] );
+
+          M := AsLeftPresentation( m );
+
+          return PresentationMorphism( M, x, U );
+
+    end );
+
+end );
+
+##
+InstallGlobalFunction( ADD_RANDOM_MORPHISM_WITH_FIXED_SOURCE_RIGHT,
+    function( category )
+
+      AddRandomMorphismWithFixedSource( category,
```

Comment:
I fear this method is invalid in the case where your ring is non-commutative, since then, transposing matrices does not behave well with taking products of matrices. 

You need to write this function in the same way as you treated the case for left modules.

Reply:
Involution have the following property Inv(AB)=Inv(B)Inv(A) and it is equal to transpose only in commutative rings. This property is exactly what needed to get a well-defined morphism: Any well-defined left presentations morphism gives us a commutative diagram between and applying the involution will give us a new commutative diagram from which we can derive a morphism of right presentations.

Reply:
Okay, so I correct my argument and state that not all rings have an involution.

Reply:
One can ask a ring S if it has involution function by `IsBound( homalgTable(S)!.Involution );`. In this PR I am adding random methods only for homalg fields, ring of integers, free polynomial rings, exterior rings and their quotients. They all have Involution functions.

There is only one very unlikely to happen scenario:

If S is some other homalg ring without involution function and one adds by hand the components S!.random_element_func and S!.random_matrix_func then in the right presentations category he will get an error whenever he ask for random object or morphism.

Reply:
Ok, it is very good that there is a way to test for the existence of an involution.

The least one should do in the code is to test whether an involution exists, and only in those cases install the random methods for the right module case. However, for me, that would be unsatisfying since that would mean treating the right module case more restrictive than the left module case.  I'll leave the choice to you.

In the graded case, you also depend on the existence of an involution for right modules, so again, either test for the existence of an involution (again with the downside of unequal behaviour between left and right modules) or don't use involutions.

Reply:
Ok, then I will treat them equally and re-implement the two methods to the right presentations without the use if Involution.

Reply:
In the current status of the PR the Involution is not required anymore.